File: 프로그래머스/4/43236. 징검다리/징검다리.js
function solution(distance, rocks, n) {
    rocks.sort((a, b) => a - b);
    rocks = [0, ...rocks, distance];
    
    const gap = rocks.map((rock, index) => rocks[index + 1] - rock)
                     .slice(0, rocks.length - 1);
    
    let [left, right] = [0, distance + 1];
    
    while (left < right) {
        let sum = 0;
        let count = 0;
        const mid = ((left + right) / 2) >> 0;
        
        gap.forEach(val => {
            sum += val;
            if (sum < mid) count++;
            else sum = 0;
        });
        
        if (count > n) right = mid;
        else left = mid + 1;
    }
    
    return left - 1;
}

File: 프로그래머스/4/43236. 징검다리/징검다리.py
def solution(distance, rocks, n):
    rocks.sort()
    rocks = [0] + rocks + [distance]
    
    gap = [rocks[i + 1] - rocks[i] for i in range(len(rocks) - 1)]

    left = 0
    right = distance + 1
    
    while left < right:
        _sum = 0
        _count = 0
        mid = (left + right) // 2
        
        for val in gap:
            _sum += val
            if _sum < mid:
                _count += 1
            else:
                _sum = 0
        
        if _count > n:
            right = mid
        else:
            left = mid + 1
    
    return left - 1